redecs_agent/redecs_agent.go

```diff
@@ -20,6 +20,95 @@ import (
 	"time"
 )
 
+const workerTimeout = 180 * time.Second
+
+type handler interface {
+	Handle(*docker.APIEvents) error
+}
+
+type dockerRouter struct {
+	handlers      map[string][]handler
+	dockerClient  *docker.Client
+	listener      chan *docker.APIEvents
+	workers       chan *worker
+	workerTimeout time.Duration
+}
+
+func dockerEventsRouter(bufferSize int, workerPoolSize int, dockerClient *docker.Client,
+	handlers map[string][]handler) (*dockerRouter, error) {
+	workers := make(chan *worker, workerPoolSize)
+	for i := 0; i < workerPoolSize; i++ {
+		workers <- &worker{}
+	}
+
+	dockerRouter := &dockerRouter{
+		handlers:      handlers,
+		dockerClient:  dockerClient,
+		listener:      make(chan *docker.APIEvents, bufferSize),
+		workers:       workers,
+		workerTimeout: workerTimeout,
+	}
+
+	return dockerRouter, nil
+}
+
+func (e *dockerRouter) start() error {
+	go e.manageEvents()
+	return e.dockerClient.AddEventListener(e.listener)
+}
+
+func (e *dockerRouter) stop() error {
+	if e.listener == nil {
+		return nil
+	}
+	return e.dockerClient.RemoveEventListener(e.listener)
+}
+
+func (e *dockerRouter) manageEvents() {
+	for {
+		event := <-e.listener
+		timer := time.NewTimer(e.workerTimeout)
+		gotWorker := false
+		// Wait until we get a free worker or a timeout
+		// there is a limit in the number of concurrent events managed by workers to avoid resource exhaustion
+		// so we wait until we have a free worker or a timeout occurs
+		for !gotWorker {
+			select {
+			case w := <-e.workers:
+				if !timer.Stop() {
+					<-timer.C
+				}
+				go w.doWork(event, e)
+				gotWorker = true
+			case <-timer.C:
+				log.Infof("Timed out waiting.")
+			}
+		}
+	}
+}
+
+type worker struct{}
+
+func (w *worker) doWork(event *docker.APIEvents, e *dockerRouter) {
+	defer func() { e.workers <- w }()
+	if handlers, ok := e.handlers[event.Status]; ok {
+		log.Infof("Processing event: %#v", event)
+		for _, handler := range handlers {
+			if err := handler.Handle(event); err != nil {
+				log.Errorf("Error processing event %#v. Error: %v", event, err)
+			}
+		}
+	}
+}
+
+type dockerHandler struct {
+	handlerFunc func(event *docker.APIEvents) error
+}
```

Comment:
can say `var _ handler = &dockerHandler{}` to statically assert dockerHandler implements handler.